src/misc/coloring.c

```diff
@@ -129,6 +131,105 @@ static igraph_error_t igraph_i_vertex_coloring_greedy_cn(const igraph_t *graph,
     return IGRAPH_SUCCESS;
 }
 
+/* DSatur coloring heuristics */
+
+static igraph_integer_t igraph_i_dsatur_select_node(
+    const igraph_t *graph, const igraph_vector_int_t *colors,
+    const igraph_vector_int_t *saturation_degree, const igraph_vector_int_t *edge_degree,
+    igraph_integer_t vc
+) {
+    igraph_integer_t most_saturated_node = -1, max_saturation = -1, max_degree = -1;
+
+    for (igraph_integer_t node = 0; node < vc; node++) {
+        //finding an uncolored node with max (saturation degree ,  degree)
+        if (VECTOR(*colors)[node] != -1){
+            continue;
+        }
+        igraph_integer_t saturation = VECTOR(*saturation_degree)[node];
+        igraph_integer_t degree = VECTOR(*edge_degree)[node] ;
+        if (( saturation < max_saturation) || (saturation == max_saturation && degree <= max_degree)) {
+            continue;
+        }
+        max_saturation = saturation;
+        max_degree = degree;
+        most_saturated_node = node;
+    }
+
+    return most_saturated_node;
+}
+
+static igraph_bool_t is_color_used_by_neighbour(
+    const igraph_vector_int_t *colors, igraph_integer_t color,
+    const igraph_vector_int_t *neighbours
+) {
+    igraph_integer_t nbr_cnt = igraph_vector_int_size(neighbours);
+
+    for (igraph_integer_t nbr_indx = 0; nbr_indx < nbr_cnt; nbr_indx++) {
+        igraph_integer_t nbr = VECTOR(*neighbours)[nbr_indx];
+        if (VECTOR(*colors)[nbr] == color) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+static void igraph_i_dsatur_update_saturation_degree(
+    const igraph_adjlist_t *adjlist, igraph_vector_int_t *saturation_degree,
+    igraph_vector_int_t *neighbours, igraph_vector_int_t *colors,
+    igraph_integer_t color
+) {
+    igraph_integer_t nbr_cnt = igraph_vector_int_size(neighbours);
+    for (igraph_integer_t nbr_indx = 0; nbr_indx < nbr_cnt; nbr_indx++){
+        igraph_integer_t nbr = VECTOR(*neighbours)[nbr_indx];
+        if (!is_color_used_by_neighbour(colors, color, igraph_adjlist_get(adjlist, nbr))) {
+            VECTOR(*saturation_degree)[nbr]++;
+        }
+    }
+}
+
+static igraph_error_t igraph_i_vertex_coloring_dsatur_cn(
+    const igraph_t *graph, igraph_vector_int_t *colors
+) {
+    igraph_adjlist_t adjlist;
+    IGRAPH_CHECK(igraph_adjlist_init(graph, &adjlist, IGRAPH_ALL, IGRAPH_LOOPS_TWICE, IGRAPH_MULTIPLE));
+    IGRAPH_FINALLY(igraph_adjlist_destroy, &adjlist);
+
+    igraph_vector_int_t saturation_degree;
+    IGRAPH_CHECK(igraph_vector_int_init(&saturation_degree, igraph_vcount(graph) ) );
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &saturation_degree);
+
+    igraph_vector_int_t edge_degree;
+    IGRAPH_CHECK(igraph_vector_int_init(&edge_degree, igraph_vcount(graph) ) );
+    IGRAPH_CHECK(igraph_degree(graph, &edge_degree, igraph_vss_all(), IGRAPH_ALL, IGRAPH_LOOPS ) );
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &edge_degree);
```

Comment:
If the `igraph_degree()` call fails, then we have a memory leak. To avoid this, `IGRAPH_FINALLY` should follow `igraph_vector_int_init()` _directly_. However, in this case, just use `IGRAPH_VECTOR_INT_INIT_FINALLY()`.